cloudbridge/cloud/providers/gce/resources.py

```diff
@@ -162,3 +168,321 @@ def zones(self):
+                len(firewall['allowed'][0]['ports']) == 1):
+                info['port'] = firewall['allowed'][0]['ports'][0]
+            return info
+        return info
+
+    def delete_firewall_id(self, firewall_id):
+        for firewall in self.iter_firewalls():
+            if firewall['id'] == firewall_id:
+                self._delete_firewall(firewall)
+        self._update_list_response()
+
+    def iter_firewalls(self, tag=None):
+        if self._list_response is None:
+            self._update_list_response()
+        if 'items' not in self._list_response:
+            return
+        for firewall in self._list_response['items']:
+            if 'targetTags' not in firewall or len(firewall['targetTags']) != 1:
+                continue
+            if 'allowed' not in firewall or len(firewall['allowed']) != 1:
+                continue
+            if tag is None or firewall['targetTags'][0] == tag:
+                yield firewall
+
+    def _delete_firewall(self, firewall):
+        project_name = self._provider.project_name
+        try:
+            response = (self._provider.gce_compute
+                                      .firewalls()
+                                      .delete(project=project_name,
+                                              firewall=firewall['name'])
+                                      .execute())
+            self._provider.wait_for_global_operation(response)
+            # TODO: process the response and handle errors.
+            return True
+        except:
+            return False
+
+    def _update_list_response(self):
+        self._list_response = (
+                self._provider.gce_compute
+                              .firewalls()
+                              .list(project=self._provider.project_name)
+                              .execute())
+
+    def _check_list_in_dict(self, dictionary, field_name, value):
+        if field_name not in dictionary:
+            return value is None
+        if (value is None or
+            len(dictionary[field_name]) != 1 or
+            dictionary[field_name][0] != value):
+            return False
+        return True
+
+
+class GCESecurityGroup(BaseSecurityGroup):
+
+    def __init__(self, delegate, tag, description=None):
+        super(GCESecurityGroup, self).__init__(delegate.provider, tag)
+        self._description = description
+        self._delegate = delegate
+
+    @property
+    def id(self):
+        return GCEFirewallsDelegate.tag_id(self._security_group)
+
+    @property
+    def name(self):
+        return self._security_group
+
+    @property
+    def description(self):
+        if self._description is not None:
+            return self._description
+        for firewall in self._delegate.iter_firewalls(self._security_group):
+            if 'description' in firewall:
+                return firewall['description']
+        return None
+
+    @property
+    def rules(self):
```

Comment:
Yeah, it's a tossup... _boto_ has an `update` method for most of the objects so it's possible to explicitly refresh an object. We don't. @nuwang I remember we had a discussion on this update topic and you were of the opinion we didn't need an `update` method. Do we just require fetching the object again to refresh it or should we implement an implicit refresh within each function? Our AWS implementation for example also does not refresh the object but returns a stale one of a change has been done outside CloudBridge.